md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2007 @@
+
+The input to execution of a function consists of:
+ - The function to be executed.
+ - The incoming argument values, one for each argument [type] of the function.
+ - A *recursion depth* value.
+
+For the duration of the execution of a function body, several data structures
+are created:
+ - A *control-flow stack*, which holds [labels] for reference from branch
+   instructions.
+ - A *value stack*, which carries values between instructions.
+ - A *locals* array, containing an element for each argument of the function in
+   the argument's [type], as well as an element for each local declaration in
+   the function.
+ - A *current position*.
+
+> Implementations need not create a literal array to store the locals, or
+literal stacks to manage values at runtime.
+
+#### Function Execution Initialization
+
+The current position starts at the first instruction in the function body. The
+value stack begins empty. The control-flow stack begins with an entry holding a
+[label] bound to the last instruction in the instruction sequence.
+
+The value of each incoming argument is copied to the local with the
+corresponding index, and the rest of the locals are initialized to all-zeros
+bit-pattern values.
+
+#### Function Body Execution
+
+The instruction at the current position is remembered, and the current position
+is incremented to point to the position following it. Then the remembered
+instruction is executed as follows.
+
+For each operand [type] in the instruction's signature in reverse order, a value
+is popped from the value stack and provided as the corresponding operand value.
+The instruction is then executed as described in the
+[Instructions Section](#instructions) entry describing it. Each of the
+instruction's return values are then pushed onto the value stack.
+
+If the current position is now past the end of the sequence,
+[function return execution](#function-return-execution) is initiated and
+execution of the function is thereafter complete.
+
+Otherwise, [execution](#function-body-execution) is resumed.
+
+#### Labels
+
+A label is a value which is either *unbound*, or *bound* to a specific position.
+
+#### Nondeterminism
+
+When semantics are specified as *nondeterministic*, a WebAssembly implementation
+may perform any one of the specified alternatives.
+
+> There is no requirement that a given implementation make the same choice every
+time, even within the same program.
+
+#### Instruction Traps
+
+Instructions may *trap*, in which case execution of the current module is
+immediately terminated and abnormal termination is reported to the embedding
+environment.
+
+#### Function Return Execution
+
+One value for each return [type] in the function signature in reverse order is
+popped from the value stack. If the function execution was prompted by a call
+instruction, these values are provided as the call's return value. Otherwise,
+they are provided to the embedding environment.
+
```

Comment:
Right, so the tailing result values could be dropped. `call $fn_returning_two_values, drop, return` would return one value, the first.